### ISSUE TITLE
BASELINE: how did pytorch quantize parameters?

### ISSUE THREAD

**Issue by ZechenM** [ZechenM/distributed-ml-training#14]
Initially I thought if I have floating-point number such as 0.0009765625, what quantization does is that it can truncate the number to only 8-bit floating-point precision to be 0.00097656

Later on, I realized on the binary level, if the binary representation floating point numbers follows IEEE 754, then doing the described quantization above would either have no help at all or add extraneous bits in the mantissa section

The question is: how did pytorch do this? What is their underlying implementation?

**Comment by ZechenM** [comment]
will figure this out and have pytorch quantization as the baseline approach

**Comment by ZixiQu** [comment]
Is the answer reduce 32bit float to 16bit?

**Comment by ZechenM** [comment]
so far that is what I found to be the most straightforward way to only quantize gradients in our setting

there are lots of quantization methods that are quantizing the whole model

there is also another quantization method called quantize_per_tensor, but I am having a little trouble playing around the scale parameter (my assumption does not align with the result)

**Comment by ZixiQu** [comment]
"so far that is what I found to be the most straightforward way to only quantize gradients in our setting. there are lots of quantization methods that are quantizing the whole model"
We don't need to quantize the whole model as we don't send models, we only send gradients. In the futher we might need to sync the model sometimes so we send the whole model, but way less frequently then gradient communication. Quantize gradient is sufficient.

"there is also another quantization method called quantize_per_tensor", it's the same thing, perhaps quantize the whole gradient in an efficient way? Maybe quantize in parallel, ours is one by one, so pretty slow.

"but I am having a little trouble playing around the scale parameter (my assumption does not align with the result)"
how is 0.1 not working? (I assume you used 0.1 and find problem)

**Comment by ZechenM** [comment]
I used 1e-8 since I found the gradient is roughly in the range of [-1e-5 , 1e-5]. I think what scale does is it divides every tensor with the scaling factor so that each value in a tensor becomes an integer. 

I printed out the gradients after quantize_per_tensor with scale = 1e-8, but they are all still floating point numbers

**Comment by ZechenM** [comment]
What they did is just they use different IEEE precisions standard:

64-bit: IEEE double-precision
32-bit: IEEE single-precision
16-bit: IEEE half-precision
